src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+            double v = A[d - 1] * A[d - 1];
+            for (auto i = d - 1; i > 0 && c * ps[i] >= 1; i--) {
+              // fix this
+              r -= ps[i];
+              c = r / (var_target - v);
+              v += A[i - 1 + indptr[rid]];
+            }
+          } else {
+            for (auto i = d - 1; i > 0 && c * ps[i] >= 1; i--) {
+              std::pop_heap(ps, ps + i + 1);
+              r -= ps[i];
+              c = r / (var_target - (d - i));
+            }
+          }
+        } else {
+          do {
+            var_1 = 0;
+            if (weights) {
+              for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+                var_1 += A[j] * A[j] / std::min(ONE, c * ps[j - indptr[rid]]);
+            } else {
+              for (auto j = indptr[rid]; j < indptr[rid + 1]; j++)
+                var_1 += ONE / std::min(ONE, c * ps[j - indptr[rid]]);
+            }
+
+            c *= var_1 / var_target;
+          } while (std::min(var_1, var_target) / std::max(var_1, var_target) < 1 - eps);
+        }
+
+        cs[i] = c;
+      }
+
+      if (!weights || iters) {
+        double cur_ex_nodes = 0;
+        for (auto it : hop_map)
+          cur_ex_nodes += std::min((FloatType)1, it.second);
+        if (cur_ex_nodes / prev_ex_nodes >= 1 - eps)
+          break;
+        prev_ex_nodes = cur_ex_nodes;
+      }
+    }
+  }
+
+  // When run distributed, random_seed needs to be equivalent.
+  auto seeds = static_cast<const int64_t *>(random_seed->data);
+  const auto num_seeds = random_seed->shape[0];
+  auto cnts = static_cast<const int64_t *>(cnt->data);
+  const auto l = (cnts[0] % cnts[1]) * 1.0 / cnts[1];
+  const auto pi = std::acos(-1.0);
+  const auto a0 = std::cos(pi * l / 2);
+  const auto a1 = std::sin(pi * l / 2);
+
+  const pcg32 ng0(seeds[0]), ng1(seeds[num_seeds - 1]);
+  std::uniform_real_distribution<FloatType> uni;
+  std::normal_distribution<FloatType> norm;
+
+  // compute number of edges first and store randoms
+  IdxType num_edges = 0;
+  phmap::flat_hash_map<IdxType, FloatType> rand_map;
+  auto rands_arr = NDArray::Empty({hop_size},
+                                  DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                  ctx);
+  FloatType* rands = static_cast<FloatType*>(rands_arr->data);
+  for (int64_t i = 0, off = 0; i < num_rows; i++) {
+    const IdxType rid = rows_data[i];
+    const auto c = cs[i];
+    for (auto j = indptr[rid]; j < indptr[rid + 1]; j++) {
+      const auto v = indices[j];
+      const auto u = nids ? nids[v] : v;
+      auto itb = rand_map.emplace(u, 0);
+      if (itb.second) {
+        auto ng = ng0;
+        ng.discard(u);
+        if (num_seeds > 1) {
+          norm.reset();
+          auto rnd = a0 * norm(ng);
+          ng = ng1;
+          ng.discard(u);
+          norm.reset();
+          rnd += a1 * norm(ng);
+          itb.first->second = std::erfc(-rnd * (FloatType)/*M_SQRT1_2*/0.70710678118654752440) / 2;
```

Comment:
What is this block doing?

Reply:
It interpolates between two random seeds smoothly so that batch_dependency works.